### PR TITLE
Add op compatibility graph coloring option

### PR DESCRIPTION
**Pull request by dmanatunga**
This feature allows the user to view the graph with the new
compatibility coloring option. This option will color the nodes on the
graph by whether they are compatible for their assigned Device. This
feature is only viewable based on the debuggerData flag.

Currently, if an op is not specified a device, the device is defaulted
to the TPU. Furthermore, all ops are considered compatible for CPU and
GPU devices, while a whitelist of compatible ops are specifed for the
TPU.

### CODE REVIEW

File: tensorboard/plugins/graph/tf_graph_common/hierarchy.ts [hunk 1]
```diff
@@ -501,6 +550,37 @@ function addNodes(h: Hierarchy, graph: SlimGraph) {
...
+            (parent.compatibilityHistogram.incompatible || 0) + 1;
+      }
+
+      // Increment capability counts for in and out embeddings
+      _.each(node.inEmbeddings, (inNode) => {
+        if (inNode.compatible) {
+          parent.compatibilityHistogram.compatible =
+              (parent.compatibilityHistogram.compatible || 0) + 1;
+        } else {
+          parent.compatibilityHistogram.incompatible =
```
Comment (chihuahua) [inline tensorboard/plugins/graph/tf_graph_common/hierarchy.ts]: I wonder if it's possible to encapsulate this logic within a function within this file, ie 

```
updateCompatibilityHistogram(node: Node) {
  if (node.compatible) {
  } else {
  }
}
```

File: tensorboard/plugins/graph/tf_graph_common/node.ts [hunk 2]
```diff
@@ -653,6 +653,53 @@ export function getFillForNode(templateIndex, colorBy,
...
+        let numCompat = node.compatibilityHistogram.compatible;
+        let numIncompat = node.compatibilityHistogram.incompatible
+
+        if (numCompat == 0 && numIncompat == 0) {
+          // Return the hue for unknown device.
+          return colorParams.UNKNOWN;
+        }
+
+        let id = "op-compat-" + node.name;
+        let escapedId = tf.graph.util.escapeQuerySelector(id);
```
Comment (chihuahua) [inline tensorboard/plugins/graph/tf_graph_common/node.ts]: I was wondering why the demo worked despite how node names could contain characters disallowed in IDs. I think that would explain it - so this is great. An alternative is to use a data atttribute, ie

    .attr('data-node-name', node.name)

and then

    let linearGradient = gradientDefs.select('linearGradient#[data-node-name=' + node.name + ']');

File: tensorboard/plugins/graph/tf_graph_common/render.ts [hunk 3]
```diff
@@ -22,7 +22,8 @@ export type Point = {x: number, y: number};
 /**
  * Color parameters for op nodes.
  */
-export let OpNodeColors = {DEFAULT_FILL: 'white', DEFAULT_STROKE: '#b2b2b2'};
+export let OpNodeColors = {DEFAULT_FILL: 'white', DEFAULT_STROKE: '#b2b2b2',
```
Comment (chihuahua) [inline tensorboard/plugins/graph/tf_graph_common/render.ts]: nit: `#ffffff` for consistency.

File: tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html [hunk 4]
```diff
@@ -83,7 +84,8 @@
         out-graph="{{_graph}}"
         out-stats="{{_stats}}"
         progress="{{_progress}}"
-out-hierarchy-params="{{_hierarchyParams}}"
+        out-hierarchy-params="{{_hierarchyParams}}"
+        op-compat-feature-enabled="[[debuggerDataEnabled]]"
```
Comment (chihuahua) [inline tensorboard/plugins/graph/tf_graph_dashboard/tf-graph-dashboard.html]: I like this feature. What if we just enabled it by default? Just like how coloring by XLA clustering is enabled by default.

`debuggerDataEnabled` is only true if the debugger plugin backend detects events issued by the debugger. That boolean is no longer guarded by a URL GET parameter, so it is a bit irrelevant to this feature.

**Comment by chihuahua** [comment]
Nice work navigating the graph logic.

File: tensorboard/plugins/graph/tf_graph_common/node.ts (hunk 2 above)
Comment (dmanatunga) [inline tensorboard/plugins/graph/tf_graph_common/node.ts]: Any preference for this, or should I keep the original method. I can't take credit though, it is based off the coloring code for ColorBy.DEVICE.

Comment (chihuahua) [inline tensorboard/plugins/graph/tf_graph_common/node.ts]: Come to think, lets keep using IDs. Nothing else is using the ID.

**Pull request merged**